### PR TITLE
Implement ast::AstNode for NameLike and move it to node_ext

### PR DESCRIPTION
**Pull request by Veykril**
With this `search`(and 2 other modules) don't necessarily go through 3 calls of `find_node_at_offset_with_descend` to find the correct node. Also makes the code that searches for NameLikes a bit easier on the eyes imo, though that can be fixed with just a helper function as well so its not that relevant.

### CODE REVIEW

File: crates/syntax/src/ast/node_ext.rs [hunk 1]
```diff
@@ -297,6 +297,52 @@ impl ast::RecordExprField {
...
+mod __ {
+    use super::{
+        ast::{Lifetime, Name, NameRef},
+        NameLike,
+    };
+    stdx::impl_from!(NameRef, Name, Lifetime for NameLike);
+}
+
 #[derive(Debug, Clone, PartialEq)]
 pub enum NameOrNameRef {
```
Comment (matklad) [inline crates/syntax/src/ast/node_ext.rs:347]: Hm, are there usages of NameOrNameRef which should be replaced by NameLike?

Comment (Veykril) [reply crates/syntax/src/ast/node_ext.rs]: The only thing that uses `NameOrNameRef` is `ast::RecordPatField::field_name`, replacing it with NameLike would look like this [`e1dbf43` (#7699)](https://github.com/rust-analyzer/rust-analyzer/pull/7699/commits/e1dbf43cf85f84c3a7e40f9731fc1f7ac96f8979), assuming keeping both isn't worth it.

Comment (matklad) [reply crates/syntax/src/ast/node_ext.rs]: ah right, field shorthand is where we need exactly this...

bors r+

Comment (matklad) [reply crates/syntax/src/ast/node_ext.rs]: oh, wait, this is with them merged, lets keep the split bors r-

**Comment by matklad** [comment]
argh, I am bad with bors, @Veykril could you add nameornameref back?

**Comment by Veykril** [comment]
was about to say I'm going to in a bit, my bad for pushing on the same branch

**Pull request merged**